akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala

```diff
@@ -4,46 +4,136 @@
+  val completionHandler = new CompletionHandler[Integer, Try[Int] ⇒ Unit] {
 
-    val ref = materializer.actorOf(context, props.withDispatcher(dispatcher))
+    override def completed(result: Integer, attachment: Try[Int] ⇒ Unit): Unit = {
+      attachment(Success(result))
+    }
 
-    (akka.stream.actor.ActorPublisher[ByteString](ref), ioResultPromise.future)
+    override def failed(ex: Throwable, attachment: Try[Int] ⇒ Unit): Unit = {
+      attachment(Failure(ex))
+    }
   }
+}
 
-  override protected def newInstance(shape: SourceShape[ByteString]): SourceModule[ByteString, Future[IOResult]] =
-    new FileSource(f, chunkSize, attributes, shape)
+/**
+ * INTERNAL API
+ * Creates simple asynchronous Source backed by the given file.
+ */
+private[akka] final class FileSource(path: Path, chunkSize: Int)
+  extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
+  require(chunkSize > 0, "chunkSize must be greater than 0")
 
-  override def withAttributes(attr: Attributes): Module =
-    new FileSource(f, chunkSize, attr, amendShape(attr))
+  val out = Outlet[ByteString]("FileSource.out")
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.fileSource
+
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val ioResultPromise = Promise[IOResult]()
+
+    val logic = new GraphStageLogic(shape) with OutHandler {
+      handler ⇒
+      val buffer = ByteBuffer.allocate(chunkSize)
+      val maxReadAhead = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+      var channel: FileChannel = _
+      var position = 0L
+      var chunkCallback: Try[Int] ⇒ Unit = _
+      var eofEncountered = false
+      var availableChunks: List[ByteString] = List.empty[ByteString]
+
+      setHandler(out, this)
+
+      override def preStart(): Unit = {
+        try {
+          // this is a bit weird but required to keep existing semantics
+          require(Files.exists(path), s"Path '$path' does not exist")
+          require(Files.isRegularFile(path), s"Path '$path' is not a regular file")
+          require(Files.isReadable(path), s"Missing read permission for '$path'")
+
+          channel = FileChannel.open(path, StandardOpenOption.READ)
+        } catch {
+          case ex: Exception ⇒
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+            throw ex
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (availableChunks.size < maxReadAhead && !eofEncountered)
+          availableChunks = readAhead(maxReadAhead, availableChunks)
+
+        if (availableChunks.nonEmpty) {
+          emitMultiple(out, availableChunks.toIterator,
+            () ⇒ if (eofEncountered) success() else setHandler(out, handler)
+          )
+          availableChunks = List.empty[ByteString]
```

Comment:
so we are waiting for next `onPull` before next `readAhead`? Might be alright, or is it worth to do `readAhead` as early as possible?